contracts/middleware/MiddlewareRemove.sol

```diff
@@ -0,0 +1,128 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.24;
+
+import {IPoolManager} from "@uniswap/v4-core/src/interfaces/IPoolManager.sol";
+import {BalanceDelta, BalanceDeltaLibrary} from "@uniswap/v4-core/src/types/BalanceDelta.sol";
+import {PoolKey} from "@uniswap/v4-core/src/types/PoolKey.sol";
+import {Proxy} from "@openzeppelin/contracts/proxy/Proxy.sol";
+import {BaseRemove} from "./BaseRemove.sol";
+import {BaseHook} from "../BaseHook.sol";
+import {Hooks} from "@uniswap/v4-core/src/libraries/Hooks.sol";
+import {IHooks} from "@uniswap/v4-core/src/interfaces/IHooks.sol";
+import {CustomRevert} from "@uniswap/v4-core/src/libraries/CustomRevert.sol";
+import {NonZeroDeltaCount} from "@uniswap/v4-core/src/libraries/NonZeroDeltaCount.sol";
+import {TransientStateLibrary} from "@uniswap/v4-core/src/libraries/TransientStateLibrary.sol";
+
+contract MiddlewareRemove is BaseRemove {
+    using CustomRevert for bytes4;
+    using Hooks for IHooks;
+    using TransientStateLibrary for IPoolManager;
+
+    error HookModifiedDeltasBeforeRemove();
+    error HookTookTooMuchFee();
+    error HookInvalidDeltasAfterRemove();
+    error MaxFeeBipsTooHigh();
+
+    constructor(IPoolManager _manager, address _impl, uint256 _maxFeeBips) BaseRemove(_manager, _impl) {
+        if (_maxFeeBips > MAX_BIPS) revert MaxFeeBipsTooHigh();
+        maxFeeBips = _maxFeeBips;
+    }
+
+    function afterRemoveLiquidity(
+        address sender,
+        PoolKey calldata key,
+        IPoolManager.ModifyLiquidityParams calldata params,
+        BalanceDelta delta,
+        bytes calldata hookData
+    ) external returns (bytes4, BalanceDelta) {
+        if (bytes32(hookData) == OVERRIDE_BYTES) {
+            (, bytes memory returnData) = address(implementation).delegatecall(
+                abi.encodeWithSelector(this.afterRemoveLiquidity.selector, sender, key, params, delta, hookData[32:])
+            );
+            return abi.decode(returnData, (bytes4, BalanceDelta));
+        }
+        (bool success, bytes memory returnData) = address(this).delegatecall{gas: GAS_LIMIT}(
```

Comment:
theres a compiler warning here bc you've already declared a returnData var in this func